### PR TITLE
Replace com_apt_itude_rules_pip with com_github_ali5h_rules_pip

### PR DESCRIPTION
**Pull request by jacobbogdanov**
We haven't really been maintaining our rules_pip, and it has fallen
behind the latest bazel development. ali5h's rules_pip offers a few
big benefits that we've been missing in i95, namely incremental fetching
of packages. It is also very compatible with the way we currently pin
packages because it also is based on pip-tools!

So that's the _why_, now onto the _what_.

First there are two patches on top of `com_github_ali5h_rules_pip`.
- `export-all.patch`: This exports all the files in the pip package
  so that we can gain access to the data files of the wheel. This is not
  the most graceful solution because it exposes _everything_ about the
  unpacked wheel, not just the data file contents. We might be able to
  come up with a more clever solution, but this will do for now.

- `sys-platform.patch`: This is a true hack. Seeing as bazel's platform
  constraint system is still in its infancy (read: not usable,) instead
  of downloaded linux wheels on macos or vice versa this simply skips
  that step. In its place is an empty `BUILD` file, so if you try and
  reference the wrong platform you'll get some opaque error about
  "no such label exists."

`com_github_ali5h_rules_pip` has a `//src:compile` target which is
identical to what `com_apt_itude_rules_pip` used internally. I
copied `compile_pip_requirements_wrapper_template.sh` and `compile.bzl`
from `com_apt_itude_rules_pip` and swapped the underlying script
to be `@com_github_ali5h_rules_pip//src:compile`. This means that
besides having a different path, the `compile_pip_requirements` rule
will continue to work exactly like it did before.

And finally the most complicated part, `pip_aliases`. First some
necessary background.

`com_github_ali5h_rules_pip` expects users to use a `requirement` macro
to map pypi package name to bazel name. Something like
```py
load("@pip3//:requirements.bzl", "requirement")

py_library(
    name = "yamlizer",
    srcs = ["yamlizer.py"],
    deps = [
        requirement("PyYAML"),
    ],
)
```
This is awesome because users don't need to know the internal mapping
and can use the same name as in the `requiements.in`. It's also the
pattern that many other projects in the space have taken.

`com_apt_itude_rules_pip` took a different approach. Instead it forces
the callers to know the translation, which is
```py
def safe_name(name):
    return name.lower().replace("-", "_").replace(".", "_")
```
So the above example would be
```py
py_library(
    name = "yamlizer",
    srcs = ["yamlizer.py"],
    deps = [
        "@pip3//pyyaml",
    ],
)
```
which isn't necessarily a bad thing, but is just different.

We also have `pip_aliases` which allows us to use `@pip2and3` and
depending on the `python_version` of the `py_binary` that depends on it
becomes the correct label. The old `pip_aliases` made assumptions about
the naming scheme from `com_apt_itude_rules_pip`, which makes sense, but
is no longer valid.

`com_github_ali5h_rules_pip` also doesn't have true cross platform
support, and the most relevant issue is
https://github.com/ali5h/rules_pip/issues/48 which explains how to
properly download linux pypi packages on macos... but now how to switch
the version based on the target/host platform.

So now that the stage is set...

`pip_aliases` has now been taught to proxy
`com_apt_itude_rules_pip`-style names to repositories of
`com_github_ali5h_rules_pip`! Which means that we don't need to update
_all_ of our pypi dependencies to use the `requirement` macro, but we
can reap the benefit of incremental fetching!

This solution gives us _both_ cross-platform building, and `@pip2and3`
support! :tada:

The way it works is quite simple (although arriving at this solution
wasn't as simple.) Here's a sudo layout of the output-base:
```
external/
  pip3/  # Created using `pip_aliases`
    WORKSPACE
    BUILD
    requirement.bzl  # Contains a macro `requirement` that proxies to sub-packages (1)
    pyyaml/
      BUILD  # Contains an `alias` from "pyyaml" to the source repo pyyaml (2)
  pip3_linux/  # Created using `pip_import`
    WORKSPACE
    BUILD
    requirements.bzl  # Contains the source `requirement` macro
  pypi_linux_36_pyyaml_5_1_2/  # Implicitly created from `pip3_linux`
    WORKSPACE
    BUILD
    pyyaml/  # Contains the unpacked wheel
```

I'm only showing one platform and one python version, but there are
currently two python versions and two platforms.

(1)
```py
def requirement(name):
  return "//" + _safe_name(name)
```
This makes it so that `load`ing this repos `requirement` macro forwards
it to the sub-package in the same repo which has the final mapping.

(2)
```py
load("@pip3_linux//:requirements.bzl", py3_linux_requirements = "requirements")
load("@pip3_osx//:requirements.bzl", py3_osx_requirements = "requirements")

alias(
    name = "pyyaml",
    actual = select({
        "platform:linux": py3_linux_requirements("PyYaml"),
        "platform:osx": py3_osx_requirements("PyYaml"),
    }),
)
```
This way we `select` the correct repo's `requirement` macro. We need to
use the macro, otherwise we wouldn't be able to get the final package
name without re-inventing `com_github_ali5h_rules_pip`.

As far as `@pip2and3`, it uses a combination of platform and python
version which is achieved by using skylib's `config_setting_group`.

Okay, that was a lot. It's turtles all the way down.

TLDR: incremental fetching, zero changes to end-developers.

### CODE REVIEW

**Review by jamesesilvia — APPROVED** [review]

**Review by ian-h-chamberlain — APPROVED** [review]
I don't think I have anything blocking, mostly just questions + comments

File: thirdparty/pip/install.bzl [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+load("@pip2_osx//:requirements.bzl", pip2_osx_install = "pip_install")
+load("@pip3_linux//:requirements.bzl", pip3_linux_install = "pip_install")
+load("@pip2_linux//:requirements.bzl", pip2_linux_install = "pip_install")
+
+def pip_install():
+    """Install all pip dependencies."""
+    pip3_osx_install()
+    pip2_osx_install()
+    pip3_linux_install()
+    pip2_linux_install()
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply thirdparty/pip/install.bzl:13]: correct, although I've never tried doing something like
```py
for py in ("2", "3"):
  load("@pip%s//:install.bzl" % py, "pip_install")
  pip_install()
```
I don't think we can do anything that clever... but maybe I'm too pessimistic

File: python/pip.bzl [hunk 2]
```diff
@@ -20,9 +22,12 @@ alias(
 )
 """.strip()
 
+def _safe_name(name):
+    return name.lower().replace("-", "_")
+
 def _parse_requirements(file_contents):
     return [
-        line.split(" ")[0].replace("-", "_")
+        line.split(" ")[0]
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply python/pip.bzl:30]: lol... `re` in starlark; https://github.com/google/starlark-go/issues/241

It's totally a hack, even in `rules_128tech` we have
`thirdparty/pip/2/requirements.in`
```
pytest<4.0
```
so it's a good thing we've never tried `pip_aliases` on that!

This definitely could be improved, even with the limited ability of https://docs.bazel.build/versions/master/skylark/lib/string.html

File: python/compile/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,8 @@
+exports_files(["compile_pip_requirements_wrapper_template.sh"])
+
+filegroup(
+    name = "compile.zip",
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply python/compile/BUILD.bazel:4]: here's where I got my info https://github.com/bazelbuild/bazel/issues/3530

File: thirdparty/pip/sys-platform.patch [hunk 4]
```diff
@@ -138,14 +138,19 @@ index 1a4f294..91cf4dd 100644
...
  
  
 +def _create_dummy_build_file(directory, sys_platform):
 +    result = """
 +# This pypi package will only appear on platform "{sys_platform}"
 +# but the current platform is "{platform}"
++py_library(
++    name = "pkg",
++
++)
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply thirdparty/pip/sys-platform.patch:152]: can you put a `fail` in a BUILD file? We could come up with a clever error message, `com_github_ali5h_rules_pip` does something like that, if you ask for a dependency it doesn't have it returns
`dep + "_not_specified_in_requirements"` or something like that.

We may or may not be able to do something similar, I added this when I was still believed that we could make `bazel fetch //...` work across the board... but that's no longer true

File: python/compile/compile.bzl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+compile_pip_requirements = rule(
+    implementation = _compile_pip_requirements_impl,
+    attrs = {
+        "requirements_in": attr.label(
+            allow_single_file = [".in"],
+            mandatory = True,
+        ),
+        "data": attr.label_list(allow_files = True),
+        "requirements_txt": attr.string(default = "requirements.txt"),
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply python/compile/compile.bzl:54]: I guess alex thought that there was no particular reason to make it mandatory and force users to end the file with `.txt`, even though we always do it that way

File: python/compile/compile.bzl [hunk 6]
```diff
@@ -0,0 +1,69 @@
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+def _get_path_relative_to_workspace(path, ctx):
+    if paths.is_absolute(path):
+        return paths.relativize(path, "/")
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply python/compile/compile.bzl:5]: this is true system `/`. I don't see a `paths.absolute` as usually rules should be relative the the workspace. In this case it's to pass to pip-compile where to write the output file so absolute is easier I guess?

File: python/pip.bzl [hunk 7]
```diff
@@ -42,20 +47,40 @@ def _generate_root_build_file(repository_ctx):
...
+def _generate_requirements_bzl(repository_ctx):
+    repository_ctx.file(
+        "requirements.bzl",
+        content = """
+def requirement(name):
+    \"\"\"Proxy the request for the requirement to the sub-package.\"\"\"
+    return "//" + _safe_name(name)
+
+def _safe_name(name):
+    return name.lower().replace("-", "_")
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply python/pip.bzl:59]: the visibility/dependency order is why I chose to just stick this in each `requirement.bzl`. Also currently these are unused... One of the tangents I went down was using `pip_aliases` _and_ switching all callers to use the `requirement` macro. If that was the case then we would be able to do
```py
load("@pip2and3//:requirement.bzl", "requirement")

...
    deps = [
        requirement("pyyaml"),
    ]
```

If we ever choose to go that route then I'm sure we'll need to revisit this

File: python/pip.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply python/pip.bzl:30]: ahh boo I thought this code was python, I guess whatever works

File: thirdparty/pip/sys-platform.patch (hunk 4 above)
[the remark this answers is not in the document]
Comment (ian-h-chamberlain) [reply thirdparty/pip/sys-platform.patch:152]: AFAIK yes, you can just put a bare `fail()` in a BUILD.bazel file, so maybe that's a good ooption

File: thirdparty/pip/export-all.patch [hunk 8]
```diff
@@ -0,0 +1,13 @@
+diff --git a/src/whl.py b/src/whl.py
+index c28dc7e..1a4f294 100644
+--- a/src/whl.py
++++ b/src/whl.py
+@@ -244,6 +244,8 @@ py_library(
+         {dependencies}
+     ],
+ )
++
++exports_files(glob(["**/**"]))
```
[the remark this answers is not in the document]
Comment (jacobbogdanov) [reply thirdparty/pip/export-all.patch:10]: this is for the `copy_file` so that we can extract data files

**Comment by jacobbogdanov** [comment]
@ian-h-chamberlain I'm going to address your comments in a follow up. Thanks for the thorough review 👍
